pkg/compactor/block_upload_test.go

```diff
@@ -1259,12 +1237,123 @@ func TestMultitenantCompactor_HandleBlockUpload_Complete(t *testing.T) {
 			default:
 				assert.Equal(t, http.StatusOK, resp.StatusCode)
 				assert.Empty(t, string(body))
+				exists, err := bkt.Exists(context.Background(), path.Join(tc.blockID, block.MetaFilename))
+				require.NoError(t, err)
+				require.True(t, exists)
 			}
+		})
+	}
+}
 
-			bkt.AssertExpectations(t)
+func TestMultitenantCompactor_ValidateAndComplete(t *testing.T) {
+	const tenantID = "test"
+	const blockID = "01G3FZ0JWJYJC0ZM6Y9778P6KD"
+	injectedError := fmt.Errorf("injected error")
 
-			if tc.verifyUpload != nil {
-				tc.verifyUpload(t, &bkt, tc.expMeta)
+	uploadingMetaPath := path.Join(tenantID, blockID, uploadingMetaFilename)
+	validationPath := path.Join(tenantID, blockID, validationFilename)
+	metaPath := path.Join(tenantID, blockID, block.MetaFilename)
+
+	validationSucceeds := func(_ context.Context) error { return nil }
+
+	testCases := []struct {
+		name                        string
+		errorInjector               func(op bucket.Operation, name string) error
+		validation                  func(context.Context) error
+		expectNoValidationFile      bool
```

Comment:
Nit: can we not use negative in the naming? I know it helps with default value, but I think being explicit about what we expect to see -- instead of "not seeing" -- would make tests easier to follow.

Reply:
Sure, I can defer to being explicit here to help the readability. This test goes through a lot of paths and I probably leaned too far into minimizing it while writing it.